Alias TagHelper#class_names to #token_list

seanpdoyle opened the pull request:
Rename the new `TagHelper#class_names` method to `TagHelper#token_list`,
and make the original available as an alias.

Inspired by a discussion on [rails/rails#40121][]

Once rendered by a browser, [an element's `class` attribute is
represented as a `DOMTokenList`][mdn-domtokenlist]. It's a
space-delimited list with unique elements.

Outside of the HTML `[class]` attribute, there are other attributes that
are represented as space-delimited lists. A non-exhaustive list
includes:

* [aria-labelledby][mdn-aria-labelledby]
* [aria-describedby][mdn-aria-describedby]
* [data-controller][data-controller] (in StimulusJS applications)
* [data-action][data-action] (in StimulusJS applications)

To support these attributes, ensure the collection of values from the
`class_names` helper is a unique Set.

[rails/rails#40121]: https://github.com/rails/rails/pull/40121#issuecomment-683366103
[mdn-domtokenlist]: https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList
[class_names]: https://edgeapi.rubyonrails.org/classes/ActionView/Helpers/TagHelper.html#method-i-class_names
[mdn-aria-labelledby]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-labelledby_attribute#Value
[mdn-aria-describedby]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute#Value
[data-action]: https://stimulusjs.org/reference/actions
[data-controller]: https://stimulusjs.org/reference/controllers

seanpdoyle committed:
Alias TagHelper#class_names to #token_list

Rename the new `TagHelper#class_names` method to `TagHelper#token_list`,
and make the original available as an alias.

Inspired by a discussion on [rails/rails#40121][]

Once rendered by a browser, [an element's `class` attribute is
represented as a `DOMTokenList`][mdn-domtokenlist]. It's a
space-delimited list with unique elements.

Outside of the HTML `[class]` attribute, there are other attributes that
are represented as space-delimited lists. A non-exhaustive list
includes:

* [aria-labelledby][mdn-aria-labelledby]
* [aria-describedby][mdn-aria-describedby]
* [data-controller][data-controller] (in StimulusJS applications)
* [data-action][data-action] (in StimulusJS applications)

To support these attributes, ensure the collection of values from the
`class_names` helper is a unique Set.

[rails/rails#40121]: https://github.com/rails/rails/pull/40121#issuecomment-683366103
[class_names]: https://edgeapi.rubyonrails.org/classes/ActionView/Helpers/TagHelper.html#method-i-class_names
[mdn-aria-labelledby]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-labelledby_attribute#Value
[mdn-aria-describedby]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute#Value
[data-action]: https://stimulusjs.org/reference/actions
[data-controller]: https://stimulusjs.org/reference/controllers

seanpdoyle commented on actionview/lib/action_view/helpers/tag_helper.rb:
```diff
@@ -295,13 +319,18 @@ def content_tag(name, content_or_options_with_block = nil, options = nil, escape
-        safe_join(build_tag_values(*args), " ")
+        TokenList.new(self, build_tag_values(*args))
       end
+      alias_method :token_list, :class_names
```
@georgeclaghorn would an alias for `tokens` make sense here too?

kaspth replied:
I wish we'd just remove class_names entirely and go with token_list as the name.

kaspth replied:
Scratch that, since class_names hasn't been released yet, we should do that.

seanpdoyle replied:
@kaspth I've pushed up https://github.com/rails/rails/pull/40146/commits/ffd18391a5fe14ea71a378437584e7b6675967e3 to abandon the alias and rename the original.

seanpdoyle replied:
@kaspth I've reverted the re-name. Do you feel that the `dom_token_list` alias is still viable?

composerinteralia commented on actionview/lib/action_view/helpers/tag_helper.rb:
```diff
@@ -121,6 +121,26 @@ def method_missing(called, *args, **options, &block)
+        end
+
+        def to_str
+          @template.safe_join([nil, to_a], " ")
```
It wasn't immediately obvious to me why we needed `nil` here (I ran the tests to discover that it is so we will get the leading space when we do `"foo" + class_names("bar", "baz")`). What do you think about:

```suggestion
          " #{to_s}"
```

seanpdoyle replied:
👏 phew, much better.

composerinteralia reviewed (commented):
Neat!

seanpdoyle commented:
@composerinteralia @georgeclaghorn while I was somewhat proud of the clever interface of `TokenList`, I've reverted that commit in favor of a much simpler https://github.com/rails/rails/pull/40146/commits/7b8d9d9fb344968bbac30942c359de6feb80f985 to declare the alias and to ensure the items in the list are unique:

---

Instead of making the values returned by `class_names`
concatenation-aware, start smaller and only alias the method name and
ensure that the resulting items in the list are unique.

The biggest motivator for this behavior is merging token lists provided
as helper arguments or partial local assigns.

For instance, consider an `application/text_field` partial:

```html+erb
<%# app/views/application/_text_field.html.erb %>
<%= form.text_field(
  method,
  class: "input input--text",
  data: { action: "blur->form-field#validate" }
) %>

<%# app/views/posts/_form.html.erb %>
<%= form_with(model: post) do |form| %>
  <%= render("text_field", {
    form: form,
    method: :name,
  }) %>
<% end %>
```

In order to pass additional `class` and `data-action` values, we could
rely on String interpolation of values we delete out of the
`local_assigns` hash, then splat out the rest of those options as
arguments:

```html+erb
<%# app/views/application/_text_field.html.erb %>
<%= form.text_field(
  method,
  class: "input input--text #{local_assigns.delete(:class)}",
  data: {
    action: "blur->form-field#validate #{local_assigns.fetch(:data, {}).delete(:action)}"
  },
  **local_assigns.except(:form, :method)
) %>

<%# app/views/posts/_form.html.erb %>
<%= form_with(model: post) do |form| %>
  <%= render("text_field", {
    form: form,
    method: :name,
    disabled: post.locked?,
    class: "input--special",
    data: {
      action: "input->form-field#somethingSpecial",
    }
  }) %>
<% end %>
```

Alternatively, we could use `token_list` (an alias for `class_names`):

```html+erb
<%= form.text_field(
  method,
  class: token_list(
    "input input--text",
    local_assigns.delete(:class)
  ),
  data: {
    action: token_list(
      "blur->form-field#validate",
      local_assigns.fetch(:data, {}).delete(:action)
    )
  },
  **local_assigns.except(:form, :method)
) %>

<%# ... %>
<%= form_with(model: post) do |form| %>
  <%= render("text_field", {
    form: form,
    method: :name,
    disabled: post.locked?,
    class: "input--special",
    data: {
      action: "input->form-field#somethingSpecial",
    }
  }) %>
<% end %>
```

This is still an awkward process, but String interpolation syntax can be
replaced by method invocation.

Comparing these two versions, concatenation doesn't seem to be a means
of making this _less_ awkward, so it feels more appropriate to start
small with a `class_names` -> `token_list` alias.

KapilSachdev commented on actionview/CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+*   Alias `ActionView::Helpers::TagHelper#class_names` to `token_list` and
+    ensure the resulting Token list items are unique.
```
```suggestion
    ensure the resulting items are unique.
```

As TokenList class is dropped, might as well drop from here too.

p8 commented:
Is it desired behaviour for `token_list` to not render a key if the value is false?
```
token_list({ foo: true, bar: false })
# => "foo"
```

seanpdoyle commented on actionview/test/template/tag_helper_test.rb:
```diff
@@ -339,15 +339,21 @@ def test_tag_builder_with_unescaped_conditional_hash_classes
+      helper = ->(*arguments) { public_send(helper_method, *arguments) }
+
+      assert_equal "song play", helper.(["song", { "play": true }])
+      assert_equal "song", helper.({ "song": true, "play": false })
```
@p8 does this line cover the behavior you're describing in https://github.com/rails/rails/pull/40146#issuecomment-684782080?

p8 replied:
@seanpdoyle  Yes, it is. I can see this being useful for `class_names` or values for aria attributes.
I'm just not sure it is expected behaviour for a method called `token_list`.
But looking at `DOMTokenList` maybe it is.

seanpdoyle replied:
@p8 I'm thinking of something like this:

```html+erb
<%= form_with(model: post) do |form| %>
  <%= form.text_field :title, aria: { 
    describedby: token_list(:always_present_label, post_title_error: post.errors[:title].any?) }
  } %>
<% end %>
```

p8 replied:
That does look nice! 👍

jonathanhefner commented on actionview/CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+*   Rename `ActionView::Helpers::TagHelper#class_names` to `token_list` and
+    ensure the resulting Token list items are unique.
+
+        token_list("foo", "foo bar")
+        # => "foo bar"
+
+    *Sean Doyle*
```
Since `class_names` hasn't been released yet, I think it might do to update the [existing CHANGELOG entry](https://github.com/rails/rails/blame/7af59e16a255da5aee2863af0eefaa9f0cfd857d/actionview/CHANGELOG.md#L46-L48) instead.  Or to remove the existing entry in favor of yours.  Either way, make sure you and everyone else gets credit! :smiley:

seanpdoyle replied:
@jonathanhefner I've pushed up https://github.com/rails/rails/pull/40146/commits/dd8dc2984a9f077eaa852411f616b7710db94fd2 to bump their entry to the top of the list.

p8 commented:
It's probably courteous to let @joelhawksley in on this proposed name change, as he is the original author of the `class_names` helper. 🙂

seanpdoyle commented:
> as he is the original author of the class_names helper. 🙂

Good point, @p8. I'm curious if this should be split into two separate PRs: 1) add the alias and uniqueness behavior, 2) finalize the renaming and remove the alias.

Would that provided a better venue to discuss the removal of `class_names`?

p8 commented:
@seanpdoyle I'm all for splitting this in two seperate PR's  👍 but I also think we should ship it.

joelhawksley commented:
> @p8 Is it desired behaviour for token_list to not render a key if the value is false?

Yes, that was my intention.

——

Thanks for reaching out @seanpdoyle! I’m excited to see the class_names code getting further attention, and that its functionality has broader applications than just building CSS class attribute strings :heart:

My first reaction seeing the proposed name change is that I have no idea what a “token” is in the context of HTML. While it’s clear after reading the PR, and may become more widely known after we ship this change as part of 6.1, I don’t find the name `token_list` to be as intuitive as `class_names`. 

What do you think of any of these names?

`#value_list`
`#values`

That being said, I don’t think `token_list` is a _bad_ name, and would be fine with sticking to it. Just my two cents.

seanpdoyle commented:
> I don’t find the name `token_list` to be as intuitive as `class_names`.

That's a good point. In context, `class_names` makes a lot more sense.

 The fact that a [`class` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/class) is accessible through a [`classList` property](https://developer.mozilla.org/en-US/docs/Web/API/Element/classList) on an Element instance which returns an instance of a [`DOMTokenList` instance](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList) might be too mired down in implementation details.

With that being said, that fact that `class_names` has value for other attributes makes the method name too restrictive, and potentially confusing when used outside its original intention.

> `#value_list`
> `#values`

Those feel a little too general purpose to me, and I'll bet there are many `values` partial-local variables that will be in conflict once upgraded.

Is `dom_token_list` an improvement? Its name resembles [`dom_id`](https://api.rubyonrails.org/classes/ActionView/RecordIdentifier.html#method-i-dom_id) and [`dom_class`](https://api.rubyonrails.org/classes/ActionView/RecordIdentifier.html#method-i-dom_class), and it communicates the actual Browser concept at play.

seanpdoyle commented:
It's also worth noting that I'm a fan of implementing the logic as `dom_token_list`, and providing a `class_names` alias, if that's the best compromise.

p8 commented:
Or maybe `dom_list`, `dom_values` or `dom_tokens`.

p8 commented:
I think this helper is probably too specialized to use outside of dom elements.
If the name is too generic it might be confusing.

joelhawksley commented:
@p8 @seanpdoyle my preference would be, in this order:

1) `dom_values`
2) `dom_token_list`,
3) `token_list`

I do think that a single method name would be a good place to start instead of providing an alias.

Whatever we pick, folks will pick it up and familiarize themselves. Please don't consider my concern here at all blocking ❤️

joelhawksley commented on actionview/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+*   Alias `ActionView::Helpers::TagHelper#class_names` to `dom_token_list` and
+    ensure the resulting Token list items are unique.
```
Since this is replacing the existing change log entry, how about something along these lines?

```suggestion
*  Add `dom_token_list` helper to create token lists with conditional values.
```

seanpdoyle commented:
The longer I sit with my thoughts on these changes, the more I prefer declaring `dom_token_list` (or whatever name we settled on) as an alias. 

I think `class_names` is a good fit for the original intention matches (mimics?) [a popular NPM package interface](https://www.npmjs.com/package/classnames).

kaspth commented:
Let's use the name token_list and make class_names the alias. Then I'll merge.
